### PR TITLE
[FLOC-4087] Implement `dvol init` in go

### PR DESCRIPTION
**Pull request by zubron**
This change implements `dvol init` functionality in our golang port. It also includes adding the golang build to our travis CI configuration.

In this change, we have added command line parsing to handle `init`. The existing python tests which do not currently pass with the golang version have been marked as `TODO`. Also added some golang unit tests which check for valid volume names as this was not covered by the existing python tests.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -18,15 +18,17 @@ cache:
 
 before_install:
   # Get packaging tools that can use and cache wheels.
-  - "pip install --upgrade pip>=7"
-  - "pip install wheel"
+  - "sudo pip install --upgrade pip>=7"
+  - "sudo pip install wheel"
```
Comment (brendan-donegan) [inline .travis.yml]: What's the reason sudo was added here?

File: cmd/init.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package cmd
```
Comment (brendan-donegan) [inline cmd/init.go]: Does this file have a purpose?

File: datalayer/main.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package datalayer
```
Comment (brendan-donegan) [inline datalayer/main.go]: Wrong place for this but I can't see how to comment on the name of a file. I think it's more typical to name a file after the package it provides. This is just from looking at some other Golang code I'm familiar with. So call this datalayer.go (and datalayer_test.go)

File: datalayer/main_test.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        if ValidVolumeName(bad) {
+            t.Error(bad + " is not a valid volume name, but it passed ValidVolumeName")
+        }
+    }
+    for _, good := range supposedGood {
+        if !ValidVolumeName(good) {
+            t.Error(good + " is a valid volume name, but it failed ValidVolumeName")
+        }
+    }
+}
```
Comment (brendan-donegan) [inline datalayer/main_test.go]: Would be good to have tests for the other functions too. Did we make any attempt to integrate coverage reports into the Travis build?

File: dvol_python/test_dvol.py [hunk 5]
```diff
@@ -95,14 +112,18 @@ def test_create_volume(self):
     def test_create_volume_already_exists(self):
         dvol = VoluminousOptions()
         dvol.parseOptions(ARGS + ["-p", self.tmpdir.path, "init", "foo"])
+        expected_output = "Error: volume foo already exists"
         try:
             dvol.parseOptions(ARGS + ["-p", self.tmpdir.path, "init", "foo"])
             # TODO assert exit code != 0
             self.assertTrue(dvol.voluminous.getOutput()[-1].strip().endswith(
```
Comment (brendan-donegan) [inline dvol_python/test_dvol.py]: This should really be assertEqual as you're comparing things. I understand this is an issue with the existing test so it's probably ok not to fix it now.

File: dvol_python/test_dvol.py [hunk 6]
```diff
@@ -132,11 +155,15 @@ def test_commit_volume(self):
...
         self.assertTrue(commit.child("file.txt").exists())
         self.assertEqual(commit.child("file.txt").getContent(), "hello!")
+    if TEST_DVOL_BINARY:
+        test_commit_volume.todo = "not expected to work in go version"
 
     def test_list_empty_volumes(self):
         dvol = VoluminousOptions()
         dvol.parseOptions(ARGS + ["-p", self.tmpdir.path, "list"])
         self.assertEqual(dvol.voluminous.getOutput(), ["  VOLUME   BRANCH   CONTAINERS "])
+    if TEST_DVOL_BINARY:
```
Comment (brendan-donegan) [inline dvol_python/test_dvol.py]: unittest's skipIf decorator seems a cleaner way to do this, not sure if trial has an equivalent decorator for TODO:

File: plugin/main.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package plugin
```
Comment (brendan-donegan) [inline plugin/main.go]: Same comment above about naming after package. Also curious if all these empty files are needed?

File: datalayer/main.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+func VolumeExists(basePath string, volumeName string) bool {
+	volumePath := filepath.FromSlash(basePath + "/" + volumeName)
+	_, err := os.Stat(volumePath)
+	return err == nil
+}
+
+func CreateVolume(basePath string, volumeName string) error {
+	volumePath := filepath.FromSlash(basePath + "/" + volumeName)
+	// TODO Factor this into a data layer object.
+	os.MkdirAll(volumePath, 0777) // XXX SEC
```
Comment (mdevilliers) [inline datalayer/main.go]: os.MkdirAll returns an error - please check it.

File: datalayer/main.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	return nil
+}
+func CreateVariant(basePath, volumeName, variantName string) error {
+	// XXX Variants are meant to be tagged commits???
+	variantPath := filepath.FromSlash(basePath + "/" + volumeName + "/branches/master")
+	os.MkdirAll(variantPath, 0777) // XXX SEC
```
Comment (mdevilliers) [inline datalayer/main.go]: os.MkdirAll returns an error - please check it.

File: datalayer/main_test.go (hunk 4 above)
Comment (mdevilliers) [inline datalayer/main_test.go]: I've added a coverage report on the golang.

**Comment by mdevilliers** [comment]
I've made some comments about checking errors if/when they are returned. apart from that looks pretty good. 

There is a tool that I can add to the pre-flight that will automate this checking and treat ignoring these as a build error - https://github.com/kisielk/errcheck I'll add a ticket to add it in the future.

**Pull request merged**